cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ControlDag.java

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.planner;
+
+import co.cask.cdap.etl.proto.Connection;
+import com.google.common.base.Joiner;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.HashMultiset;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multiset;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.UUID;
+
+/**
+ * A DAG (directed acyclic graph) where edges represent a happens-before relationship.
+ * In these types of scenarios, certain edges may be redundant and can be removed.
+ * This simplifies the dag into something that is much easier to work with if it needs
+ * to be used as a fork-join type of dag for workflow execution.
+ */
+public class ControlDag extends Dag {
+  private static final Set<String> EMPTY = ImmutableSet.of();
+  private final Multiset<String> nodeVisits;
+
+  public ControlDag(Collection<Connection> connections) {
+    super(connections);
+    this.nodeVisits = HashMultiset.create();
+  }
+
+  /**
+   * Record that this node was visited.
+   *
+   * @param node node that was visited
+   * @return the number of times this node was visited, including the visit from this call
+   */
+  public int visit(String node) {
```

Comment:
i should have left it out of this pr, but it will be used during workflow configuration in order to know when to call join()